Yoshi/QAKit/EnvironmentMenu/YoshiEncodableEnvironmentManager.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  YoshiEncodableEnvironmentManager.swift
+//  Yoshi
+//
+//  Created by Kanglei Fang on 25/09/2017.
+//  Copyright © 2017 Prolific Interactive. All rights reserved.
+//
+
+import Foundation
+
+/// Environment selection event.
+public typealias EnvironmentChangeEvent = (Environment) -> Void
+
+/// A YoshiEnvironmentManager that persisit user's environment selection using NSUserDefaults.
+open class YoshiEncodableEnvironmentManager: YoshiEnvironmentManager {
+    
+    public var currentEnvironment: Environment {
+        didSet {
+            archive(environment: currentEnvironment)
+            onEnvironmentChange?(currentEnvironment)
+        }
+    }
+    
+    public private(set) var environments: [Environment]
+    
+    private static let YoshiEnvironemntKey = "YoshiEnvironment"
+    
+    private static let encoder = JSONEncoder()
+    private static let decoder = JSONDecoder()
+    
+    private var onEnvironmentChange: EnvironmentChangeEvent?
+    
+    /// Initialize the environement manager with the given environments.
+    /// The manger will retrieve the archaved environment selection from NSUserDefaults if any.
+    /// Othereise the first environment will be used as the current environment.
+    ///
+    /// - Parameters:
+    ///   - environments: Available Environments.
+    ///   - onEnvironmentChange: Callback when environment is changed.
+    ///                          Notice that the callback will be invoked when the manager is initialized.
+    public init(environments: [Environment], onEnvironmentChange: EnvironmentChangeEvent?) {
+        self.environments = environments
+        self.onEnvironmentChange = onEnvironmentChange
+        if let archivedEnvironment = YoshiEncodableEnvironmentManager.archivedEnvironment {
+            currentEnvironment = archivedEnvironment
+        } else {
+            let defaultEnvironment = environments[0]
```

Comment:
Maybe use a `precondition` or `assert` to make sure that `environments` is not empty (so that you are not unintentionally crashing on this line)?

Reply:
That provides nothing other than a comment, however, do you think it necessary to prevent a crash here by not selecting a default environment here?

Reply:
It signals to the user of the framework that you are intentional about it and I think that's important. 

As for your question, I am not quite sure if not having a default-selected environment makes much sense. Your app needs to have some sort of API environment it's making calls against.